File: Python/HackerRank_Practice_Lab/Task_Simulation/queensAttack.py

def get_total_attackable_squares(board_size, obstacle_count, queen_row, queen_column, obstacles):
    max_range = {
        'north': board_size - queen_row,
        'south': queen_row - 1,
        'west': queen_column - 1,
        'east': board_size - queen_column,
        'north_west': min(board_size - queen_row, queen_column - 1),
        'north_east': min(board_size - queen_row, board_size - queen_column),
        'south_west': min(queen_row - 1, queen_column - 1),
        'south_east': min(queen_row - 1, board_size - queen_column)
    }

    for obs_row, obs_column in obstacles:
        if obs_column == queen_column:
            if obs_row > queen_row:
                max_range['north'] = min(max_range['north'], obs_row - queen_row - 1)
            else:
                max_range['south'] = min(max_range['south'], queen_row - obs_row - 1)
        
        elif obs_row == queen_row:
            if obs_column > queen_column:
                max_range['east'] = min(max_range['east'], obs_column - queen_column - 1)
            else:
                max_range['west'] = min(max_range['west'], queen_column - obs_column - 1)
        
        elif abs(obs_row - queen_row) == abs(obs_column - queen_column):
            row_diff = obs_row - queen_row
            col_diff = obs_column - queen_column
            
            if row_diff > 0 and col_diff < 0:
                max_range['north_west'] = min(max_range['north_west'], row_diff - 1)
            elif row_diff > 0 and col_diff > 0:
                max_range['north_east'] = min(max_range['north_east'], row_diff - 1)
            elif row_diff < 0 and col_diff < 0:
                max_range['south_west'] = min(max_range['south_west'], abs(row_diff) - 1)
            elif row_diff < 0 and col_diff > 0:
                max_range['south_east'] = min(max_range['south_east'], abs(row_diff) - 1)

    return sum(max_range.values())

def main():
    first_multiple_input = input().rstrip().split()
    n = int(first_multiple_input[0])
    k = int(first_multiple_input[1])

    second_multiple_input = input().rstrip().split()
    r_q = int(second_multiple_input[0])
    c_q = int(second_multiple_input[1])

    obstacles = []
    for _ in range(k):
        obstacles.append(list(map(int, input().rstrip().split())))
    result = get_total_attackable_squares(n, k, r_q, c_q, obstacles)
    
    print(result)

if __name__ == "__main__":
    main()

File: problems/hackerrank/simulation/queensAttack/queensAttack.go
package simulation



// GetTotalAttackableSquares calculates the total squares a queen can attack.
func GetTotalAttackableSquares(n, k, rQ, cQ int, obstacles [][]int) int {
	maxRange := map[string]int{
		"north":      n - rQ,
		"south":      rQ - 1,
		"west":       cQ - 1,
		"east":       n - cQ,
		"north_west": min(n-rQ, cQ-1),
		"north_east": min(n-rQ, n-cQ),
		"south_west": min(rQ-1, cQ-1),
		"south_east": min(rQ-1, n-cQ),
	}

	for _, obs := range obstacles {
		obsR, obsC := obs[0], obs[1]
		if obsC == cQ {
			if obsR > rQ {
				maxRange["north"] = min(maxRange["north"], obsR-rQ-1)
			} else {
				maxRange["south"] = min(maxRange["south"], rQ-obsR-1)
			}
		} else if obsR == rQ {
			if obsC > cQ {
				maxRange["east"] = min(maxRange["east"], obsC-cQ-1)
			} else {
				maxRange["west"] = min(maxRange["west"], cQ-obsC-1)
			}
		} else if abs(obsR-rQ) == abs(obsC-cQ) {
			rowDiff := obsR - rQ
			colDiff := obsC - cQ
			if rowDiff > 0 && colDiff < 0 {
				maxRange["north_west"] = min(maxRange["north_west"], rowDiff-1)
			} else if rowDiff > 0 && colDiff > 0 {
				maxRange["north_east"] = min(maxRange["north_east"], rowDiff-1)
			} else if rowDiff < 0 && colDiff < 0 {
				maxRange["south_west"] = min(maxRange["south_west"], abs(rowDiff)-1)
			} else if rowDiff < 0 && colDiff > 0 {
				maxRange["south_east"] = min(maxRange["south_east"], abs(rowDiff)-1)
			}
		}
	}

	total := 0
	for _, v := range maxRange {
		total += v
	}
	return total
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}
